### PR TITLE
Implement `aesara.tensor.matmul`

### LINKED ISSUE #488: Implement `aesara.tensor.matmul`
As in https://numpy.org/doc/stable/reference/generated/numpy.matmul.html

### PR DESCRIPTION
**Pull request by zoj613**
closes #488

This implements an aesara equivalent of `np.matmul`.

The behavior depends on the arguments in the following way.

If both arguments are 2-D they are multiplied like conventional matrices.

If either argument is N-D, N > 2, it is treated as a stack of matrices residing in the last two indexes and broadcast accordingly.

If the first argument is 1-D, it is promoted to a matrix by prepending a 1 to its dimensions. After matrix multiplication the prepended 1 is removed.

If the second argument is 1-D, it is promoted to a matrix by appending a 1 to its dimensions. After matrix multiplication the appended 1 is removed.

`matmul` differs from `dot` in two important ways:

Multiplication by scalars is not allowed.

Stacks of matrices are broadcast together as if the matrices were elements, respecting the signature `(n,k),(k,m)->(n,m)`

References
---------------
https://numpy.org/doc/stable/reference/generated/numpy.matmul.html

### CODE REVIEW

**Review by brandonwillard — COMMENTED** [review]
Looks great!

The next two important methods that need to be implemented are `Op.grad` (or `Op.L_op`) and `Op.infer_shape`.  

They're both optional, but they can really make or break the usefulness of an `Op`.  At the very least, it's good to have "stubs" for those that explicitly say they aren't implemented.

You can use `tests.unittest_tools.verify_grad` to create numeric gradient tests (or even `tests.tensor.utils.makeTester` for a basic test suite), and `tests.unittest_tools.InferShapeTester` to create automate the `Op.infer_shape` tests.

File: aesara/tensor/__init__.py
Comment (brandonwillard) [inline aesara/tensor/__init__.py:115]:
```suggestion
```

File: aesara/tensor/nlinalg.py
Comment (brandonwillard) [inline aesara/tensor/nlinalg.py:847]: `broadcast_shape` is a relatively expensive thing to do just to get the number of output dimensions.

File: aesara/tensor/nlinalg.py
Comment (brandonwillard) [inline aesara/tensor/nlinalg.py:841]: `broadcastable` indicates which dimensions have shape `1`.  This light form of shape inference can be pretty important for optimizations, so it's always worth attempting to determine the `broadcastable` pattern given the `broadcastable` patterns of the inputs.

Take a look at `aesara.tensor.math.Dot.make_node` for some of the relevant logic.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Is it possible that `np.matmul` functionality can be implemented by a helper function that dispatches to the existing `aesara.tensor.math.Dot` and `aesara.tensor.math.tensordot`?  If so, that would save considerable time and effort attempting to (re)implement the `Op.grad` and `Op.infer_shape` logic.

File: tests/tensor/test_nlinalg.py
Comment (brandonwillard) [inline tests/tensor/test_nlinalg.py:127]: The tests are currently failing under `aesara.config.floatX = "float32"`.  This is due to the NumPy test values having type `float64` by default.  If you convert the values with `.astype(config.floatX)`, the problem should be solved.

**Comment by zoj613** [comment]
> Is it possible that `np.matmul` functionality can be implemented by a helper function that dispatches to the existing `aesara.tensor.math.Dot` and `aesara.tensor.math.tensordot`? If so, that would save considerable time and effort attempting to (re)implement the `Op.grad` and `Op.infer_shape` logic.

Not sure this is best given that `matmul` behaviour is different from dot's depending on the shape of the inputs. See the PR description. I managed to implement the `infer_shape` method without much hassle.

File: aesara/tensor/nlinalg.py
Comment (zoj613) [reply aesara/tensor/nlinalg.py]: You're right, i didn't think that one through. Resolved in latest commit.

**Comment by zoj613** [comment]
> `tests.unittest_tools.InferShapeTester` to create automate the `Op.infer_shape` tests.

Could you elaborate on this part? Do I necessarily have to implement a `test_infer_shape` method or is this automated by inheriting from `InferShapeTester`? I see that some tests implement this method and some dont. not sure what is the best practice here.

**Comment by brandonwillard** [comment]
> > `tests.unittest_tools.InferShapeTester` to create automate the `Op.infer_shape` tests.
> 
> Could you elaborate on this part? Do I necessarily have to implement a `test_infer_shape` method or is this automated by inheriting from `InferShapeTester`? I see that some tests implement this method and some dont. not sure what is the best practice here.

Yeah, you need to create a subclass and then call the method(s) it provides.  It can be a useful tool, but, if you want to make custom tests, that's also fine.

**Comment by zoj613** [comment]
> > > `tests.unittest_tools.InferShapeTester` to create automate the `Op.infer_shape` tests.
> > 
> > 
> > Could you elaborate on this part? Do I necessarily have to implement a `test_infer_shape` method or is this automated by inheriting from `InferShapeTester`? I see that some tests implement this method and some dont. not sure what is the best practice here.
> 
> Yeah, you need to create a subclass and then call the method(s) it provides. It can be a useful tool, but, if you want to make custom tests, that's also fine.

as in `_compile_and_check`?

File: aesara/tensor/nlinalg.py
Comment (rlouf) [inline aesara/tensor/nlinalg.py:856]: I think it would be better to be "dumb" (and explicit) about the `if... elif... else` logic here.

File: aesara/tensor/nlinalg.py
Comment (brandonwillard) [reply aesara/tensor/nlinalg.py:833]: As a means of confirming that the shapes are compatible with the operation, but in the context of partial static shape information (i.e. the old `TensorType.broadcastable`, and now the new `TensorType.shape`)?

If that's what you're asking, then&mdash;generally speaking&mdash;yeah, we consider every opportunity to do things like that, especially when it helps prevent more cryptic errors down the line.

If we're talking about a dot product operation, then we should be able to check that the last dimension of `a` has the same broadcast value as the first dimension of `b` (i.e. `a.type.broadcastable[-1] == b.type.broadcastable[0]`).  We could do even better now that `TensorType.shape` is available.

File: aesara/tensor/nlinalg.py
Comment (brandonwillard) [reply aesara/tensor/nlinalg.py:874]: It might work, but this explicit case handling will definitely produce simpler graphs in most cases.  Hopefully, we'll come up with an efficient solution to canonicalizing the results of `Op.infer_shape`, but, at this point, we don't have one implemented, so local "optimizations" like this aren't unwelcome.

In summary, if the output can be represented that succinctly it really should be, but due to other limitations of this library, it's currently acceptable.

Comment (zoj613) [reply aesara/tensor/nlinalg.py:874]:
```  File "/home/zoj/.conda/envs/aesara/lib/python3.7/site-packages/numpy/lib/stride_tricks.py", line 470, in broadcast_shapes
    arrays = [np.empty(x, dtype=[]) for x in args]
  File "/home/zoj/.conda/envs/aesara/lib/python3.7/site-packages/numpy/lib/stride_tricks.py", line 470, in <listcomp>
    arrays = [np.empty(x, dtype=[]) for x in args]
TypeError: 'TensorVariable' object cannot be interpreted as an integer
```
numpy is not too happy about receiving TensorVariable in the np.broadcast_shapes implementation.  Here is a reproduction example:
```python
In [21]: np.empty((at.as_tensor(3), at.as_tensor(4)), dtype=[])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-21-a80110ee1c26> in <module>
----> 1 np.empty((at.as_tensor(3), at.as_tensor(4)), dtype=[])

TypeError: 'TensorConstant' object cannot be interpreted as an integer
```

Comment (zoj613) [reply aesara/tensor/nlinalg.py:874]: but using the aesara implementation produces its own errors:
```
aesara/tensor/basic_opt.py:1008: KeyError
----------------------------------------------------------------- Captured log call ------------------------------------------------------------------
ERROR    aesara.graph.opt:opt.py:221 SeqOptimizer apply <aesara.tensor.basic_opt.ShapeOptimizer object at 0x7f21fa584310>
ERROR    aesara.graph.opt:opt.py:222 Traceback:
ERROR    aesara.graph.opt:opt.py:223 Traceback (most recent call last):
  File "/home/zoj/dev/aesara/aesara/graph/opt.py", line 279, in apply
    sub_prof = optimizer.optimize(fgraph)
  File "/home/zoj/dev/aesara/aesara/graph/opt.py", line 101, in optimize
    self.add_requirements(fgraph)
  File "/home/zoj/dev/aesara/aesara/tensor/basic_opt.py", line 1411, in add_requirements
    fgraph.attach_feature(ShapeFeature())
  File "/home/zoj/dev/aesara/aesara/graph/fg.py", line 569, in attach_feature
    attach(self)
  File "/home/zoj/dev/aesara/aesara/tensor/basic_opt.py", line 1221, in on_attach
    self.on_import(fgraph, node, reason="on_attach")
  File "/home/zoj/dev/aesara/aesara/tensor/basic_opt.py", line 1288, in on_import
    self.set_shape(r, s)
  File "/home/zoj/dev/aesara/aesara/tensor/basic_opt.py", line 1053, in set_shape
    shape_vars.append(self.unpack(s[i], r))
  File "/home/zoj/dev/aesara/aesara/tensor/basic_opt.py", line 1008, in unpack
    s_i = self.shape_of[x][i]
KeyError: Subtensor{::, ::, int64, int64}.0
```
@brandonwillard do you have an idea of what might causing this?

Comment (brandonwillard) [reply aesara/tensor/nlinalg.py:874]: > numpy is not too happy about receiving TensorVariable in the np.broadcast_shapes implementation. Here is a reproduction example:

You definitely don't want to pass NumPy Aesara variables.  Use the equivalent Aesara function instead.

Comment (zoj613) [reply aesara/tensor/nlinalg.py:874]: That doesn't work either as shown in the stack trace right above your post. Do you know why its throwing an error like that?

Comment (zoj613) [reply aesara/tensor/nlinalg.py:874]: the shape inference works if I use something like `(3, 2, 4, 5), (3, 2, 5, 7)`, but fails for `(3, 2, 4, 5), (2, 5, 7)`

Comment (brandonwillard) [reply aesara/tensor/nlinalg.py:874]: > but using the aesara implementation produces its own errors:

The error is saying that `ShapeFeature`, which keeps track of all the shapes in a `FunctionGraph`, is missing an entry for a `Variable` it's encountered.  This can happen when `ShapeFeature` comes across a `Variable` that wasn't properly added to the `FunctionGraph`, because such variables wouldn't have triggered `FunctionGraph`'s callbacks and&mdash;as a result&mdash;wouldn't have been picked up by `ShapeFeature`.

If this is caused by the use of `broadcast_shape` in `MatMul.infer_shape`, then it's probably because `broadcast_shape` is introducing those elements.  If I recall, `broadcast_shape` will use the shapes of its arguments, and any time a shape is requested for a `Variable` (e.g. 
 via the `TensorVariable.shape` method) a `Shape` `Op` is created in the resulting graph.  The function of `ShapeFeature` is to replace those `Shape`s with the results of the relevant `Op`'s `Op.infer_shape` method (like the one you're writing).

What's likely happening is that your `MatMul.infer_shape` is creating a graph that computes the shape of a `Variable`, and that graph itself contains `Shape` `Op`s that need to be converted by `ShapeFeature`, but the inputs to some of those `Shape` `Op`s are new `Variable`s that were just created inside of `MatMul.infer_shape`, so `ShapeFeature` has never seen them and raises that `KeyError`.

The solution is to always avoid using `Shape` `Op`s inside of the graphs produced by an `Op.infer_shape`.  This is part of the reason why `Op.infer_shape` has an argument that specifies the shapes of the `Op`s inputs, because those are all one should need in order to compute the shape of the `Op`'s outputs.

Unfortunately, this restriction and approach are largely due to some very old design issues.  I actually opened an issue earlier today about this very situation: #749.

Comment (zoj613) [reply aesara/tensor/nlinalg.py:874]: Never mind. I was completely unaware that `broadcast_shape` has a `arrays_are_shapes` parameter. everything works now. Embarrassing moment there *face palm*.

Comment (brandonwillard) [reply aesara/tensor/nlinalg.py]: It's nothing to be embarrassed about; the source of the confusion is primarily due to an unnecessarily confusing and convoluted shape handling system.

File: aesara/tensor/nlinalg.py
Comment (zoj613) [reply aesara/tensor/nlinalg.py:856]: you're right, i made the conditionals more explicit in latest commit.

File: aesara/tensor/nlinalg.py
Comment (zoj613) [reply aesara/tensor/nlinalg.py:833]:
```python
In [2]: a = np.random.rand(1, 2, 4, 5) 
In [3]: b = np.random.rand(1, 5, 7)
In [4]: np.matmul(a,b).shape
Out[4]: (1, 2, 4, 7)
In [9]: at.matmul(a, b).type.broadcastable
Out[9]: (False, False, False, False)
```
Just to be sure I understand @rlouf 's concern correctly, we're referring to the problem shown in the snippet above where the first element of the output's `broadcastable` is `False` when it should be `True` because of the `1`?. How would we know the shape inside of `make_node` in order to encode this information while creating the output node? I thought it was impossible?

Comment (brandonwillard) [reply aesara/tensor/nlinalg.py:833]: If I understood correctly, he wasn't necessarily talking about having the correct broadcast pattern for the `Type`s output by `MatMul` (although that's also very important), but that the broadcast patterns could be used for input validation in `MatMul.make_node`.

For example, take the following incompatible inputs:
```python
import numpy as np
import aesara.tensor as at


a = np.random.rand(1)
b = np.random.rand(2)

np.matmul(a, b)
# ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 2 is different from 1)
```

Our Aesara version of `np.matmul` should raise an error for the same inputs, as well, and it will; however, it will raise that error _at run-time_ and not when the symbolic graph is created.  Now, in this example, we might've been able to tell that the inputs were not compatible based on the partial shape information provided by the broadcast patterns:

```python
a_at = at.as_tensor(a)
b_at = at.as_tensor(b)

a_at.type.broadcastable
# (True,)
b_at.type.broadcastable
# (False,)
```
Unfortunately, this isn't really enough to work with, unless you take the broadcast patterns too strictly, because these boolean broadcast values really translate to static shape information like `a_at.type.shape == (1,)` and `b_at.type.shape == (None,)` (this was conceptually true even before #711 was merged).  The `None` indicates that we do not know the shape of `b_at` at "compile" time.  It could be _anything_, including `(1,)`, so we can't actually use it for input validation in this (and most) case(s).

Nevertheless, you seem to have found an example for which the broadcast pattern of the _output_ of `MatMul` isn't correct, and that should definitely be fixed.

Comment (rlouf) [reply aesara/tensor/nlinalg.py:833]: That's what I meant, sorry for not expressing myself more clearly.

File: aesara/tensor/nlinalg.py
Comment (rlouf) [reply aesara/tensor/nlinalg.py]: Yeah definitely nothing to be embarrassed about, the API is also very confusing here.

File: aesara/tensor/nlinalg.py
Comment (zoj613) [reply aesara/tensor/nlinalg.py]: I implemented a mechanism to validate the input shapes at compile time. If overuse of conditionals makes the code a bit off-putting so any refactoring tips are welcome.

File: aesara/tensor/nlinalg.py
Comment (brandonwillard) [inline aesara/tensor/nlinalg.py:894]: Looks like you've anticipated one of the major improvements we need to make: https://github.com/aesara-devs/aesara/issues/732!

File: tests/tensor/test_nlinalg.py
Comment (brandonwillard) [inline tests/tensor/test_nlinalg.py:131]: This use of a shared RNG state induces test method order dependence; instead, if you create and seed the RNG objects within each independent test unit, they can be run in any order (and in parallel) and produce consistent results.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: tests/tensor/test_nlinalg.py
Comment (zoj613) [reply tests/tensor/test_nlinalg.py:131]: Just to be clear: by "test unit" you mean each independent `test_*` method?

Comment (brandonwillard) [reply tests/tensor/test_nlinalg.py:131]: > Just to be clear: by "test unit" you mean each independent `test_*` method?

Yes

File: aesara/tensor/nlinalg.py
Comment (rlouf) [inline aesara/tensor/nlinalg.py:837]: As per your comment, I don't find this necessarily off-putting. If you really want to avoid repeating conditionals you can always use dictionaries:

```python
cases = [
    {
        "case_fn": lambda x1, x2: x1.ndim==1 and x2.ndim ==1,
        "error_fn": lambda x1_shape, x2_shape: x1_shape[0] != x2_shape[0],
        "error_msg": "1d inputs must have the same length."
    }
]

if validate:
    for case in cases:
        x1_shape, x2_shape = shapes
        if case["cond_fn"](x1, x2) and case["error_fn"](x1_shape, x2_shape):
            raise ValueError(case["error_msg"])      
```

But it doesn't make the code any clearer imho. Repetition is not necessarily bad. Having said that you could return before all the conditionals if `validate == False`

Comment (zoj613) [reply aesara/tensor/nlinalg.py:837]: Yeah, it looks like trying to be clever here is not going to improve readability. I'll leave it as is.

**Comment by Sayam753** [comment]
@zoj613 is there anything left to be completed in this PR? 

The PR https://github.com/aesara-devs/aesara/pull/808 is blocked to use `matmul` operation (implemented in this PR).

**Comment by brandonwillard** [comment]
It looks like this can't be rebased by maintainers, and we need to (squash and) rebase to make sure it passes with all the recent changes&mdash;especially the shape-inference-related ones.  @zoj613, is "Allow edits and access to secrets by maintainers" not enabled/checked?

**Comment by Sayam753** [comment]
@brandonwillard is this PR ready to be merged?

File: aesara/tensor/__init__.py
Comment (brandonwillard) [inline aesara/tensor/__init__.py:117]:
```suggestion
```

File: aesara/tensor/nlinalg.py
Comment (brandonwillard) [inline aesara/tensor/nlinalg.py:908]:
```suggestion
    """Compute a matrix product of two tensor variables.
```

File: aesara/tensor/nlinalg.py
Comment (brandonwillard) [inline aesara/tensor/nlinalg.py:906]: Let's add type hints here instead of the docstring.

File: aesara/tensor/nlinalg.py
Comment (brandonwillard) [inline aesara/tensor/nlinalg.py:923]:
```suggestion
        `x1`, `x2` are 1-d vectors.
```

File: aesara/tensor/nlinalg.py
Comment (brandonwillard) [inline aesara/tensor/nlinalg.py:929]:
```suggestion
        If the last dimension of `x1` is not the same size as the second-to-last
        dimension of `x2`. If a scalar value is passed in.
```

File: aesara/tensor/nlinalg.py
Comment (brandonwillard) [inline aesara/tensor/nlinalg.py:943]:
```suggestion
    `matmul` differs from `dot` in two important ways:
```

File: aesara/tensor/nlinalg.py
Comment (brandonwillard) [inline aesara/tensor/nlinalg.py:945]:
```suggestion
    - Multiplication by scalars is not allowed, use `mul` instead.
```

File: aesara/tensor/nlinalg.py
Comment (brandonwillard) [inline aesara/tensor/nlinalg.py:947]:
```suggestion
        respecting the signature ``(n, k), (k, m) -> (n, m)``:
```

File: tests/tensor/test_nlinalg.py
Comment (brandonwillard) [reply tests/tensor/test_nlinalg.py:131]: It looks like all the tests are sharing the same class-level RNG object.  As I mentioned earlier, this will make the results order-dependent.  We need to construct the RNG object within each individual test in order to avoid that.

N.B. A fixture could be used if you don't want to copy-paste the RNG construction code each time.

File: tests/tensor/test_nlinalg.py
Comment (brandonwillard) [inline tests/tensor/test_nlinalg.py:202]: Do these test arrays need to be this large in order to reach the relevant code path?

File: tests/tensor/test_nlinalg.py
Comment (brandonwillard) [inline tests/tensor/test_nlinalg.py:222]: Same here; unless we're testing functionality that requires large arrays, let's keep the sizes as low as possible.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Some minor changes are needed; otherwise, it looks good.

**Comment by purna135** [comment]
Thank you so much for this PR, @zoj613! I'm keeping an eye on it, and it appears that some minor changes are required before this can be merged. Are you able to push this PR to the boundary line? I have a strong dependency on this PR.

**Comment by brandonwillard** [comment]
I've just noticed that `numpy.matmul` is essentially a `ufunc` of `dot`.  With that in mind, the right way to do this is to first close https://github.com/aesara-devs/aesara/issues/695 and use that functionality to implement this.

**Comment by purna135** [comment]
Hello, @brandonwillard. Is this PR now ready to be merged?
I have some PR that are waiting for this PR to be merged because my work is dependent on `matrix_inverse` at #808, which is once again blocked by this PR : (

**Review by brandonwillard — APPROVED** [review]
We can merge in order to move https://github.com/aesara-devs/aesara/pull/808 along, but we need to keep the associated issue open until at least a gradient implementation is provided.  

Preferably, we should have both a gradient and non-Python implementation for the addition of an `Op`, and, in this case, https://github.com/aesara-devs/aesara/pull/757 is where our efforts need to be focused to accomplish that.

**Comment by zoj613** [comment]
I am looking into adding a grad/L_op/R_op method for this. It appears that it won't be as straight forward as re-using one for the `Dot` Op since `matmul` has slight differences in behavior, especially for dimensions > 2.

**Comment by brandonwillard** [comment]
> I am looking into adding a grad/L_op/R_op method for this. It appears that it won't be as straight forward as re-using one for the `Dot` Op since `matmul` has slight differences in behavior, especially for dimensions > 2.

Yeah, I realized that the logic for doing that is a slightly specialized form of the logic we need in https://github.com/aesara-devs/aesara/pull/757, so it's better that we focus our efforts there.

**Pull request merged**
